bindings/js/api.js

```diff
@@ -1,76 +1,150 @@
 // jshint esversion: 6
 
-var fs = require('fs');
-var path = require('path');
-var loader = require(path.resolve(__dirname, './loader.js'));
-var shell = require(path.resolve(__dirname, './shell.js'));
-var api_json = require(path.resolve(__dirname, './api.json'));
+if (typeof module !== 'undefined') {
+  var fs = require('fs');
+  var path = require('path');
+  var loader = require(path.resolve(__dirname, './loader.js'));
+  var shell = require(path.resolve(__dirname, './shell.js'));
+  var api_promise = Promise.resolve(require(path.resolve(__dirname, './api.json')));
+  var layouts_promise = Promise.resolve(require(path.resolve(__dirname, './layouts.json')));
+  var modules_promise = Promise.resolve(shell.my_modules.map(m => {
+    var source = fs.readFileSync(path.resolve(__dirname, './' + m + ".wasm"));
+    return new Uint8Array(source);
+  }));
+
+} else {
+  var loader = this;
+  var shell = this;
+  var api_promise = fetch("wasm/api.json").then(r => r.json());
+  var layouts_promise = fetch("wasm/layouts.json").then(r => r.json());
+  var modules_promise = Promise.all(my_modules.map(m => fetch("wasm/" + m + ".wasm")))
+    .then(response => Promise.all(response.map(r => r.arrayBuffer())));
+}
+
+// HELPERS FOR SIZE FIELDS
+// -----------------------
+
+// Grammar of size fields: var<OP>n where
+// - var is the name of a variable
+// - n is an integer constant
+// - <OP> is +, - or *
+// - no spaces at all.
+var parseOp = (arg, op) => {
+  let [ var_, const_ ] = arg.split(op);
+  return [ var_, op, const_ ];
+};
+
+var parseSize = (arg) => {
+  if (arg.indexOf("+") >= 0)
+    return parseOp(arg, "+");
+  if (arg.indexOf("-") >= 0)
+    return parseOp(arg, "-");
+  if (arg.indexOf("*") >= 0)
+    return parseOp(arg, "*");
+  if (arg.indexOf("/") >= 0)
+    return parseOp(arg, "/");
+  return [ arg ];
+};
+
+var evalSize = function(arg, args_int32s) {
+  let [ var_, op, const_ ] = parseSize(arg);
+  switch (op) {
+    case "+":
+      return args_int32s[var_] + parseInt(const_);
+    case "-":
+      return args_int32s[var_] - parseInt(const_);
+    case "*":
+      return args_int32s[var_] * parseInt(const_);
+    case "/":
+      let x = parseInt(const_);
+      if (args_int32s[var_] % x != 0)
+        throw new Error("Argument whose length is "+arg+" is not a multiple of "+x);
+      return args_int32s[var_] / x;
+    default:
+      return args_int32s[var_];
+  }
+};
+
+var invertSize = function(arg, total) {
```

Comment:
Given `total = len<op>size`, this returns `len` by inverting the formula?

Reply:
fixed